### PR TITLE
feat: implement /api/v1/labels for prometheus

### LINKED ISSUE #1541: Support `/api/v1/labels` from Prometheus
### What problem does the new feature solve?

The interface [`/api/v1/labels`](https://prometheus.io/docs/prometheus/latest/querying/api/#getting-label-names) returns a list of label names, which is the name of columns that semantic type is `TAG` in GreptimeDB.

Related to #1016 

### What does the feature do?

Implement [`/api/v1/labels`](https://prometheus.io/docs/prometheus/latest/querying/api/#getting-label-names) to fetch all the columns' name which semantic type is `TAG`.

### Implementation challenges

- Implement the HTTP endpoint
- Fetch and process the column names

### PR DESCRIPTION
**Pull request by haohuaijin**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement `/api/v1/labels` for prometheus. 
use a enum struct to represent `PromData` for return labels's data.
custom `Deserialize` method `Matchs` to support parsing repeated `match[]`

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

close #1541

### CODE REVIEW

File: src/servers/src/prom.rs [hunk 1]
```diff
@@ -463,6 +477,208 @@ pub async fn range_query(
...
+        }),
+    }
+}
+
+/// Retrieve labels name from record batches
+fn record_batches_to_labels_name(
+    batches: RecordBatches,
+    labels: &mut HashSet<String>,
+) -> Result<()> {
+    let mut tag_column_indices = Vec::new();
```
Comment (killme2008) [inline src/servers/src/prom.rs:623]: I think we can return all the column names as labels except `greptime_value`. It's not precise, but acceptable.

The `greptime_value` is defined at https://github.com/GreptimeTeam/greptimedb/blob/497b1f9dc987e118bbd482e0b92e36889a33db7c/src/servers/src/prometheus.rs#L35

**Review by killme2008 — COMMENTED** [review]
Please add a intergartion test in `servers/tests/http/prometheus_test.rs`

File: src/servers/src/prom.rs (hunk 1 above)
Comment (haohuaijin) [reply src/servers/src/prom.rs:623]: Do your mean we return all the column names including timestamp column name and field column name.

**Comment by haohuaijin** [comment]
Should we put the test cast to `tests-integration/tests/http.rs`, I find we test `/api/v1/query` and `/api/v1/query_range` in this file.
https://github.com/GreptimeTeam/greptimedb/blob/027707d9698006732fdd643f112fe105e8225566/tests-integration/tests/http.rs#L293-L301

File: src/servers/src/prom.rs [hunk 2]
```diff
@@ -463,6 +477,208 @@ pub async fn range_query(
     PromJsonResponse::from_query_result(result, metric_name, Some(ValueType::Matrix)).await
 }
 
+#[derive(Debug, Default, Serialize, JsonSchema)]
+struct Matches(Option<String>);
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/servers/src/prom.rs:481]: I mean that we can use `Vec<String>` to replace `Option<String>`.

File: src/servers/src/prom.rs [hunk 3]
```diff
@@ -463,6 +477,205 @@ pub async fn range_query(
...
+
+    let mut labels: HashSet<String> = HashSet::new();
+    labels.insert(METRIC_NAME.to_string());
+
+    for query in queries {
+        let prom_query = PromQuery {
+            query,
+            start: start.clone(),
+            end: end.clone(),
+            step: DEFAULT_LOOKBACK_STRING.to_string(),
```
Comment (waynexia) [inline src/servers/src/prom.rs:572]: step and lookback are two different things. But we can use its value here. Better to leave a todo comment.

File: src/servers/src/prom.rs [hunk 4]
```diff
@@ -463,6 +477,205 @@ pub async fn range_query(
...
+            .collect()
+    });
+    let Some(queries) = matches.or_else(|| {
+        form_params.matches.0.map(|s| {
+            s.split('$')
+                .filter(|s| !s.is_empty())
+                .map(|s| s.to_owned())
+                .collect::<Vec<String>>()
+        })}
+    ) else { return PromJsonResponse::error("Unsupported", "match[] parameter is required"); };
```
Comment (waynexia) [inline src/servers/src/prom.rs:549]: Format looks weird in this line, but rustfmt doesn't complain. Will rustfmt rewrite it if we add some newlines?

**Review by waynexia — COMMENTED** [review]
Thanks for your contribution. Could you add some HTTP integration tests for this new interface? You can find examples in `test-integration/tests/http.rs`

File: src/servers/src/prom.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (waynexia) [reply src/servers/src/prom.rs:481]: Theoretically this `matches` parameter is a valid json string, we can try to deserialize it from json to vec string

**Comment by killme2008** [comment]
@haohuaijin There is a conflict have to be resolved.

After discussing with @waynexia , we thought that maybe we don't need to impl `Deserialize` for `Matches`. Just define the `LabelsQuery` as below:

```rust
#[derive(Debug, Default, Serialize, Deserialize, JsonSchema)]
pub struct LabelsQuery {
    start: Option<String>,
    end: Option<String>,
    #[serde(flatten)]
    matches: Vec<String>,
    db: Option<String>,
}
```

The `serde` crate will take care of the serialization.

File: tests-integration/tests/http.rs [hunk 5]
```diff
@@ -314,6 +314,19 @@ pub async fn test_prom_http_api(store_type: StorageType) {
         .await;
     assert_eq!(res.status(), StatusCode::OK);
 
+    // labels
+    let res = client
+        .get("/api/v1/labels?match[]=up&start=0&end=100")
+        .send()
+        .await;
+    assert_eq!(res.status(), StatusCode::OK);
```
Comment (killme2008) [inline tests-integration/tests/http.rs:322]: Can we assert the response?

**Comment by haohuaijin** [comment]
> @haohuaijin There is a conflict have to be resolved.
> 
> After discussing with @waynexia , we thought that maybe we don't need to impl `Deserialize` for `Matches`. Just define the `LabelsQuery` as below:
> 
> ```rust
> #[derive(Debug, Default, Serialize, Deserialize, JsonSchema)]
> pub struct LabelsQuery {
>     start: Option<String>,
>     end: Option<String>,
>     #[serde(flatten)]
>     matches: Vec<String>,
>     db: Option<String>,
> }
> ```
> 
> The `serde` crate will take care of the serialization.

If we define struct like above, we input `match[]` should be a sequence not many `match[]=up`, and `#[serde(flatten)]` only can use in structs and maps. should we do this?

File: tests-integration/tests/http.rs (hunk 5 above)
Comment (haohuaijin) [reply tests-integration/tests/http.rs:322]: Yes, we can assert this.

File: src/servers/src/prom.rs [hunk 6]
```diff
@@ -463,6 +478,191 @@ pub async fn range_query(
...
 
+#[derive(Debug, Default, Serialize, JsonSchema)]
+struct Matches(Vec<String>);
+
+#[derive(Debug, Default, Serialize, Deserialize, JsonSchema)]
+pub struct LabelsQuery {
+    start: Option<String>,
+    end: Option<String>,
+    #[serde(flatten)]
+    matches: Matches,
```
Comment (gitccl) [inline src/servers/src/prom.rs:489]: I think we can use this?
```rust
#[serde(rename = "match[]")]
matches: Option<Vec<String>>
```

Comment (waynexia) [reply src/servers/src/prom.rs:489]: `match[]` seems to be a (array of) valid promql query. 

>maybe it also can't deserialize duplicate match[]

We don't need to handle duplicate `match[]`

File: tests-integration/tests/http.rs [hunk 7]
```diff
@@ -313,6 +313,19 @@ pub async fn test_prom_http_api(store_type: StorageType) {
         .await;
     assert_eq!(res.status(), StatusCode::OK);
 
+    // labels
+    let res = client
+        .get("/api/v1/labels?match[]=up&start=0&end=100")
```
Comment (waynexia) [inline tests-integration/tests/http.rs:318]: Could you add another case that pass multiple `match[]`?

**Review by waynexia — COMMENTED** [review]
Other parts looks good to me

File: tests-integration/tests/http.rs (hunk 7 above)
Comment (haohuaijin) [reply tests-integration/tests/http.rs]: Ok, maybe I should remove the `start` and `end`, because we have the default value.

**Review by waynexia — APPROVED** [review]

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**